File: python/graphs/numIslands.py
from typing import List
from collections import deque
class Solution:
    def numIslands2(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        rv = 0
        visited = set()


        def bfs(r,c):
            q = deque()
            visited.add((r,c))
            q.append((r,c))
            while q:
                row, col = q.popleft()
                directions = [[1,0] , [-1,0] , [0,1] , [0,-1]]
                for dr, dc in directions:
                    newRow, newCol = row + dr, col + dc
                    if ((newRow) in range(rows) and 
                        (newCol) in range(cols) and
                        grid[newRow][newCol] == '1' and 
                        (newRow, newCol) not in visited):

                        visited.add((newRow, newCol))
                        q.append((newRow, newCol))



        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1' and (r, c) not in visited:
                    bfs(r,c)
                    rv += 1
        
        return rv

    def numIslands(self, grid: List[List[str]]) -> int:
        rv = 0
        height = len(grid)
        width = len(grid[0])
        
        checked = [[False for _ in range(width)] for _ in range(height)]
        
        
        for i in range(height):
            for j in range(width):
                if grid[i][j] == '1' and checked[i][j] == False:
                    rv += 1
                    self.checkLand(grid,checked,i,j,height,width)
        return rv

    def checkLand(self, grid: List[List[str]], checked: List[List[bool]], i: int, j: int, height: int, width: int):
        q = deque([(i,j)])
        while q:
            i, j = q.popleft()
            if 0 <= i < height and 0 <= j < width and grid[i][j] == '1' and checked[i][j] == False:
                checked[i][j] = True
                q.extend([(i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)])
            


map = [['1','1'],['0','0'],['0','0']]


print(Solution().numIslands(map))
    

File: java/numIslands.java
class Solution {
    int [][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public int numIslands(char[][] grid) {
        int rv = 0;
        int num_rows = grid.length;
        int num_cols = grid[0].length;
        boolean[][] checked_grid = new boolean[num_rows][num_cols];
        for (int r = 0; r < num_rows; r++)
        {
            for (int c = 0; c < num_cols; c++)
            {
                checked_grid[r][c] = false;
            }
        }

        for (int r = 0; r < num_rows; r++)
        {
            for (int c = 0; c < num_cols; c++)
            {
                if (checked_grid[r][c]) continue;
                
                checked_grid[r][c] = true;
                if (grid[r][c] == '0') continue;
                else
                {
                    rv++;
                    grid[r][c] = '1';
                    dfs(grid, r, c, num_rows, num_cols);
                }
            }
        }
        return rv;
    }

    public void dfs(char[][] grid, int r, int c, int nr, int nc)
    {
        if ((r < 0) || (r >= nr) || (c < 0) || (c >= nc)) return;
        if (grid[r][c] == '0') return;
        grid[r][c] = '0';
        for (int[] d: directions)
        {
            dfs(grid, r + d[0], c + d[1], nr, nc);
        }
    }
}